From bf940430a833613c6237c08bdb7fd1ea325c570b Mon Sep 17 00:00:00 2001
From: Ritanya Thangavel <rthangav@qti.qualcomm.com>
Date: Wed, 3 Jun 2026 10:40:45 +0530
Subject: [PATCH] [QMESH][RDK]Support for Airties TLV extensions

Created separate TLV handler files to generate EMEX TLV's with OneWiFi support in RDK
Added support to send Radio capabilities as part of AP capability report

Change-Id: I5f38d3438efff92456b5badbb6e9cec8b9ec8bcb
Signed-off-by: Ritanya Thangavel <rthangav@qti.qualcomm.com>
---
 agent/src/beerocks/slave/agent_db.h           |    2 +-
 common/beerocks/multi_vendor/CMakeLists.txt   |    7 +-
 .../include/tlvf_airties_utils_onewifi.h      |  109 ++
 .../src/multi_vendor_airties_onewifi.cpp      |   55 +
 .../src/tlvf_airties_utils_onewifi.cpp        | 1227 +++++++++++++++++
 framework/platform/bpl/cfg/rdk/bpl_cfg.cpp    |   73 +
 framework/platform/bpl/include/bpl/bpl_cfg.h  |   16 +
 .../airties/tlvAirtiesRadioCapability.yaml    |    2 +-
 8 files changed, 1486 insertions(+), 5 deletions(-)
 create mode 100644 common/beerocks/multi_vendor/onewifi/include/tlvf_airties_utils_onewifi.h
 create mode 100644 common/beerocks/multi_vendor/onewifi/src/multi_vendor_airties_onewifi.cpp
 create mode 100644 common/beerocks/multi_vendor/onewifi/src/tlvf_airties_utils_onewifi.cpp

diff --git a/agent/src/beerocks/slave/agent_db.h b/agent/src/beerocks/slave/agent_db.h
index ceb2c3800..e75229074 100644
--- a/agent/src/beerocks/slave/agent_db.h
+++ b/agent/src/beerocks/slave/agent_db.h
@@ -420,7 +420,7 @@ public:
         // 12-byte attribute containing the MCS set as defined in 802.11ax
         std::array<uint8_t, beerocks::message::HE_MCS_SET_SIZE> he_mcs_set;
 
-        bool eht_supported           = false;
+        bool eht_supported = false; ///< Is 802.11be (Extremely High Throughput / WiFi 7) protocol supported
         bool smd_support             = false;
         bool smd_dl_data_fwd_support = false;
         typedef struct {
diff --git a/common/beerocks/multi_vendor/CMakeLists.txt b/common/beerocks/multi_vendor/CMakeLists.txt
index 30075fa6c..9c0ea63b3 100644
--- a/common/beerocks/multi_vendor/CMakeLists.txt
+++ b/common/beerocks/multi_vendor/CMakeLists.txt
@@ -28,9 +28,9 @@ if (USE_PRPLMESH_WHM)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_PRPLMESH_WHM")
     list(APPEND multi_vendor_sources ${MODULE_PATH}/airties/src/multi_vendor_airties.cpp)
     list(APPEND multi_vendor_sources ${MODULE_PATH}/airties/src/tlvf_airties_utils.cpp)
-else ()
-    list(APPEND multi_vendor_sources ${MODULE_PATH}/nl80211/src/multi_vendor_airties_nl80211.cpp)
-    list(APPEND multi_vendor_sources ${MODULE_PATH}/nl80211/src/tlvf_airties_utils_nl80211.cpp)
+elseif(TARGET_PLATFORM_TYPE STREQUAL "rdk")
+    list(APPEND multi_vendor_sources ${MODULE_PATH}/onewifi/src/multi_vendor_airties_onewifi.cpp)
+    list(APPEND multi_vendor_sources ${MODULE_PATH}/onewifi/src/tlvf_airties_utils_onewifi.cpp)
 endif()
 
 # Build the library
@@ -59,6 +59,7 @@ target_include_directories(${PROJECT_NAME}
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/airties/include>
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/qca/include>
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/nl80211/include>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/onewifi/include>
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/>
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../../agent/src/beerocks/slave/>
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../../../framework/platform>
diff --git a/common/beerocks/multi_vendor/onewifi/include/tlvf_airties_utils_onewifi.h b/common/beerocks/multi_vendor/onewifi/include/tlvf_airties_utils_onewifi.h
new file mode 100644
index 000000000..c8dac905d
--- /dev/null
+++ b/common/beerocks/multi_vendor/onewifi/include/tlvf_airties_utils_onewifi.h
@@ -0,0 +1,109 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+ * 
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef __TLVF_AIRTIES_UTILS_H__
+#define __TLVF_AIRTIES_UTILS_H__
+
+#include <tlvf/CmduMessageTx.h>
+
+#include <beerocks/tlvf/beerocks_message.h>
+#include <tlvf/airties/tlvAirtiesEthernetStats.h>
+
+/*
+ * This Macro is to enable or disable
+ * a bit in an octet. This is used for setting
+ * Gateway/Extender information in Device Info TLV.
+ */
+#define TLV_BIT_ENABLE 0x1
+#define TLV_BIT_DISABLE 0x0
+
+#define CPU_TEMP_FILE "/sys/devices/virtual/thermal/thermal_zone0/temp"
+#define MEMINFO_FILE "/proc/meminfo"
+#define STAT_FILE "/proc/stat"
+#define STAT_CPU_TXT "cpu "
+#define MEMCACHED_TXT "Cached:"
+#define MEMTOTAL_TXT "MemTotal:"
+#define MEMFREE_TXT "MemFree:"
+#define MEMBUFFER_TXT "Buffers:"
+#define STAT_IDLE_IND 3
+
+// Macros for different Link Speed Types
+#define BITRATE_10 10
+#define BITRATE_100 100
+#define BITRATE_1K 1000
+#define BITRATE_2_5K 2500
+#define BITRATE_5K 5000
+#define BITRATE_10K 10000
+
+#define BYTES_IN_KB 1024
+#define COUNTERS_SIZE 6 // Size of the counter is 6 octets.
+
+#define BCAST_BYTES_SUPPORTED (BCAST_BYTES_SENT && BCAST_BYTES_RECVD)
+#define MCAST_BYTES_SUPPORTED (MCAST_BYTES_SENT && MCAST_BYTES_RECVD)
+
+namespace airties {
+
+class tlvf_airties_utils {
+public:
+    bool is_airties_platform_common_stp_enabled() const;
+
+    /**
+     * @brief Adds a new Airties Version Reporting TLV to given message.
+     *
+     * @param[in,out] cmdu_tx CDMU message.
+     *
+     * @return True on success and false otherwise.
+     */
+    static bool add_airties_version_reporting_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_airties_deviceinfo_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_device_metrics(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_airties_msgtype_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_airties_ethernet_interface_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_airties_ethernet_stats_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_radio_capability(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_airties_ethernet_1905_neighbor_device_list_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    static bool add_airties_ethernet_non_1905_neighbor_device_list_tlv(ieee1905_1::CmduMessageTx &cmdu_tx);
+
+    /**
+     * @brief Assigns a sequential port ID for the given interface name.
+     *
+     * Each interface name is mapped to a unique, persistent ID.
+     * If the interface was already assigned an ID, the same ID is returned.
+     * Otherwise, the next available ID is assigned, starting from 1.
+     *
+     * @param interface_name The interface name used as the key (e.g., "eth0", "lan1", "wan").
+     * @return The assigned port ID for this interface.
+     */
+    static uint8_t assign_unique_port_id(const std::string &interface_name);
+
+    /**
+     * Function to convert the counter to
+     * Big Endian format, so that when sent over the transport
+     * layer, the counters wont get swapped.
+     * This workaround is needed because, non-native integers are not supported
+     * in TLVF framework as of now. Hence, a list of 6 integers is used to
+     * to store 48-bit integer. This will get swapped when sent over
+     * transport layer. To avoid this, we are converting and sending the values.
+     * This will be a temporaray workaround until complete support
+     * for non-native integer is implemented in TLVF.
+     */
+    static uint64_t swap_and_convert_counter(uint64_t val);
+
+};
+} // namespace airties
+
+#endif // __TLVF_AIRTIES_UTILS_H__
diff --git a/common/beerocks/multi_vendor/onewifi/src/multi_vendor_airties_onewifi.cpp b/common/beerocks/multi_vendor/onewifi/src/multi_vendor_airties_onewifi.cpp
new file mode 100644
index 000000000..af8c54ba1
--- /dev/null
+++ b/common/beerocks/multi_vendor/onewifi/src/multi_vendor_airties_onewifi.cpp
@@ -0,0 +1,55 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+ * 
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "multi_vendor.h"
+#include "tlvf_airties_utils_onewifi.h"
+
+using namespace airties;
+using namespace ieee1905_1;
+
+/**
+ * @class MultiVendorAirties
+ * @brief This class registers Airties vendor-specific TLV handlers for IEEE 1905.1 message types.
+ */
+class MultiVendorAirties : public multi_vendor::tlvf_handler, public airties::tlvf_airties_utils {
+public:
+    MultiVendorAirties()
+    {
+        LOG(INFO) << "MultiVendorAirties: registering Airties vendor-specific TLV handlers";
+
+        multi_vendor::tlvf_handler::register_handler(eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE,
+                                                     add_airties_version_reporting_tlv);
+        multi_vendor::tlvf_handler::register_handler(eMessageType::AP_AUTOCONFIGURATION_WSC_MESSAGE,
+                                                     add_airties_deviceinfo_tlv);
+
+        multi_vendor::tlvf_handler::register_handler(
+            eMessageType::AP_AUTOCONFIGURATION_SEARCH_MESSAGE, add_airties_version_reporting_tlv);
+
+        multi_vendor::tlvf_handler::register_handler(eMessageType::AP_CAPABILITY_REPORT_MESSAGE,
+                                                     add_airties_version_reporting_tlv);
+        multi_vendor::tlvf_handler::register_handler(eMessageType::AP_CAPABILITY_REPORT_MESSAGE,
+                                                     add_radio_capability);
+
+        multi_vendor::tlvf_handler::register_handler(eMessageType::AP_METRICS_RESPONSE_MESSAGE,
+                                                     add_device_metrics);
+        multi_vendor::tlvf_handler::register_handler(eMessageType::AP_METRICS_RESPONSE_MESSAGE,
+                                                     add_airties_ethernet_stats_tlv);
+
+        multi_vendor::tlvf_handler::register_handler(eMessageType::TOPOLOGY_RESPONSE_MESSAGE,
+                                                     add_airties_ethernet_interface_tlv);
+        multi_vendor::tlvf_handler::register_handler(eMessageType::TOPOLOGY_RESPONSE_MESSAGE,
+                                                     add_airties_ethernet_1905_neighbor_device_list_tlv);
+        multi_vendor::tlvf_handler::register_handler(eMessageType::TOPOLOGY_RESPONSE_MESSAGE,
+                                                     add_airties_ethernet_non_1905_neighbor_device_list_tlv);
+    }
+};
+
+namespace {
+MultiVendorAirties g_airties_vendor_init;
+}
diff --git a/common/beerocks/multi_vendor/onewifi/src/tlvf_airties_utils_onewifi.cpp b/common/beerocks/multi_vendor/onewifi/src/tlvf_airties_utils_onewifi.cpp
new file mode 100644
index 000000000..0b6e04adf
--- /dev/null
+++ b/common/beerocks/multi_vendor/onewifi/src/tlvf_airties_utils_onewifi.cpp
@@ -0,0 +1,1227 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+ * 
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+ 
+#include "tlvf_airties_utils_onewifi.h"
+#include "agent_db.h"
+#include <bcl/beerocks_config_file.h>
+#include <bcl/beerocks_utils.h>
+#include <bcl/son/son_wireless_utils.h>
+#include <bpl/bpl_cfg.h>
+#include <bpl/common/utils/utils.h>
+#include <cstring>
+#include <easylogging++.h>
+#include <linux/if_bridge.h>
+#include <mapf/common/utils.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <tlvf/airties/eAirtiesTLVId.h>
+#include <tlvf/airties/supported_features.h>
+#include <tlvf/airties/tlvAirtiesDeviceInfo.h>
+#include <tlvf/airties/tlvAirtiesDeviceMetrics.h>
+#include <tlvf/airties/tlvAirtiesEthernetInterface.h>
+#include <tlvf/airties/tlvAirtiesEthernetStats.h>
+#include <tlvf/airties/tlvAirtiesMsgType.h>
+#include <tlvf/airties/tlvAirtiesRadioCapability.h>
+#include <tlvf/airties/tlvVersionReporting.h>
+#include <tlvf/airties/tlvAirtiesEth1905NeighborDeviceList.h>
+#include <tlvf/airties/tlvAirtiesEthNon1905NeighborDeviceList.h>
+#include <tlvf/ieee_1905_1/sVendorOUI.h>
+#include <tlvf/common/sMacAddr.h>
+
+using namespace airties;
+using namespace beerocks;
+
+/*
+ * Enum contains the different Link Types
+ */
+enum link_types_enum {
+    UNDEFINED = 0,
+    TYPE_10MBPS,
+    TYPE_100MBPS,
+    TYPE_1GBPS,
+    TYPE_2_5GBPS,
+    TYPE_5GBPS,
+    TYPE_10GBPS
+};
+
+/*
+ * Optional Ethernet counters.
+ * 1 = supported, 0 = not supported.
+ *
+ * Some counters cannot currently be retrieved from the platform (/sys, ethtool, etc.),
+ * so they remain disabled. If we later find a valid source for any of them,
+ * simply switch the corresponding value to 1.
+ */
+enum supported_stats_enum {
+    BCAST_BYTES_SENT  = 0,
+    BCAST_BYTES_RECVD = 0,
+    BCAST_PKTS_SENT   = 0,
+    BCAST_PKTS_RECVD  = 0,
+    MCAST_BYTES_SENT  = 0,
+    MCAST_BYTES_RECVD = 0,
+    MCAST_PKTS_SENT   = 0,
+    MCAST_PKTS_RECVD  = 1
+};
+
+/*
+ * Following function returns the link type.
+ * This is based on the requirement,
+ * Bit 7:4 Supported Link Type Link type (0=undefined, 1=10Mbps, 2=100Mbps,
+                                          3=1Gbps, 4=2,5Gbps, 5=5Gpbs, 6=10Gbps)
+ * Bit 3:0 Current Link Type Link type (0=undefined, 1=10Mbps, 2=100Mbps,
+                                          3=1Gbps, 4=2,5Gbps, 5=5Gpbs, 6=10Gbps
+ */
+static uint8_t bitrate_to_link_type(uint32_t bit_rate)
+{
+    uint8_t value;
+    switch (bit_rate) {
+    case BITRATE_10:
+        value = TYPE_10MBPS;
+        break;
+    case BITRATE_100:
+        value = TYPE_100MBPS;
+        break;
+    case BITRATE_1K:
+        value = TYPE_1GBPS;
+        break;
+    case BITRATE_2_5K:
+        value = TYPE_2_5GBPS;
+        break;
+    case BITRATE_5K:
+        value = TYPE_5GBPS;
+        break;
+    case BITRATE_10K:
+        value = TYPE_10GBPS;
+        break;
+    default:
+        value = UNDEFINED;
+        LOG(INFO) << "Invalid value in Current Link Type " << bit_rate;
+        break;
+    }
+    return value;
+}
+
+/*
+ * Function to set the 2 byte supported_stats_val field in
+ * Ethernet Stats TLV. Based on the counter support available in the data model,
+ * this 2 octet field is set.
+ * Bit15: BcastBytesSentPresent
+ * Bit14: BcastBytesReceivedPresent
+ * Bit13: BcastPacketsSentPresent
+ * Bit12: BcastPacketsReceivedPresent
+ * Bit11: McastBytesSentPresent
+ * Bit10: McastBytesReceivedPresent
+ * Bit9 : McastPacketsSentPresent
+ * Bit8 : McastPacketsReceivedPresent
+ * Bit 7:0 FutureStatsPresent.
+ */
+uint16_t set_supp_stats_val()
+{
+    uint16_t var = 0x0000;
+
+    std::vector<supported_stats_enum> macros = {
+        BCAST_BYTES_SENT, BCAST_BYTES_RECVD, BCAST_PKTS_SENT, BCAST_PKTS_RECVD,
+        MCAST_BYTES_SENT, MCAST_BYTES_RECVD, MCAST_PKTS_SENT, MCAST_PKTS_RECVD};
+
+    for (size_t i = 0; i < macros.size(); ++i) {
+        var |= (macros[i] << (15 - i));
+    }
+
+    return var;
+}
+
+/*
+ * Function to add the Ethernet Interface TLV
+ * to Topology Response Message
+ */
+bool tlvf_airties_utils::add_airties_ethernet_interface_tlv(ieee1905_1::CmduMessageTx &m_cmdu_tx)
+{
+    auto tlvAirtiesEthIntf = m_cmdu_tx.addClass<airties::tlvAirtiesEthernetInterface>();
+    if (!tlvAirtiesEthIntf) {
+        LOG(ERROR) << "addClass wfa_map::tlvDeviceEthernetInterface failed";
+        return false;
+    }
+
+    tlvAirtiesEthIntf->vendor_oui() =
+        (sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES));
+    tlvAirtiesEthIntf->tlv_id() =
+        static_cast<int>(airties::eAirtiesTlVId::AIRTIES_ETHERNET_INTERFACE);
+
+    auto lan_ifaces = beerocks::net::network_utils::linux_get_lan_interfaces();
+    for (const auto &lan_iface : lan_ifaces) {
+        std::string iface_mac;
+
+        if (!beerocks::net::network_utils::linux_iface_get_mac(lan_iface, iface_mac)) {
+            continue;
+        }
+
+        uint32_t link_speed;
+        uint32_t max_speed;
+        bool is_full_duplex;
+
+        if (!net::network_utils::linux_iface_get_link_settings(lan_iface, link_speed, max_speed,
+                                                               is_full_duplex)) {
+            LOG(WARNING) << "Failed to get link settings for " << lan_iface;
+            continue;
+        }
+
+        auto interface_list = tlvAirtiesEthIntf->create_interface_list();
+
+        interface_list->port_id() = airties::tlvf_airties_utils::assign_unique_port_id(lan_iface);
+        interface_list->eth_mac() = tlvf::mac_from_string(iface_mac);
+        interface_list->set_eth_intf_name(lan_iface);
+
+        interface_list->flags1().eth_port_admin_state =
+            beerocks::net::network_utils::linux_iface_is_up(lan_iface);
+        interface_list->flags1().eth_port_link_state =
+            beerocks::net::network_utils::linux_iface_is_up_and_running(lan_iface);
+        interface_list->flags1().eth_port_duplex_mode = is_full_duplex ? 1 : 0;
+        interface_list->flags2().supported_link_type  = bitrate_to_link_type(max_speed);
+        interface_list->flags2().current_link_type    = bitrate_to_link_type(link_speed);
+
+        tlvAirtiesEthIntf->add_interface_list(interface_list);
+    }
+    LOG(INFO) << "Added Ethernet Interface TLV";
+    return true;
+}
+
+uint64_t tlvf_airties_utils::swap_and_convert_counter(uint64_t val)
+{
+    val &= 0xFFFFFFFFFFFF;
+
+    uint64_t byte1 = (val & 0x0000000000FF) << 40;
+    uint64_t byte2 = (val & 0x00000000FF00) << 24;
+    uint64_t byte3 = (val & 0x000000FF0000) << 8;
+    uint64_t byte4 = (val & 0x0000FF000000) >> 8;
+    uint64_t byte5 = (val & 0x00FF00000000) >> 24;
+    uint64_t byte6 = (val & 0xFF0000000000) >> 40;
+
+    uint64_t swapped = (byte1 | byte2 | byte3 | byte4 | byte5 | byte6);
+
+    return swapped;
+}
+
+/*
+ * Function to return byte value converted
+ * to KB value.
+ */
+uint64_t convertBytes_to_Kb(uint64_t bytes_val) { return (bytes_val / BYTES_IN_KB); }
+
+inline void insert_ethernet_stats_item(std::shared_ptr<airties::cPortList> &port_list,
+                                       uint64_t counter)
+{
+    auto statsItem = port_list->create_statsItem();
+    if (!statsItem) {
+        LOG(ERROR) << "Failed to create stats item!";
+        return;
+    }
+
+    uint64_t swapped = airties::tlvf_airties_utils::swap_and_convert_counter(counter);
+
+    statsItem->set_item(&swapped, COUNTERS_SIZE);
+
+    port_list->add_statsItem(statsItem);
+}
+
+/*
+ * Function to add the Ethernet Statistics TLV
+ * to AP Metrics Response Message
+ */
+bool tlvf_airties_utils::add_airties_ethernet_stats_tlv(ieee1905_1::CmduMessageTx &m_cmdu_tx)
+{
+    auto tlvAirtiesEthStats = m_cmdu_tx.addClass<airties::tlvAirtiesEthernetStats>();
+    if (!tlvAirtiesEthStats) {
+        LOG(ERROR) << "addClass wfa_map::tlvAirtiesEthStats failed";
+        return false;
+    }
+
+    tlvAirtiesEthStats->vendor_oui() =
+        (sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES));
+    tlvAirtiesEthStats->tlv_id() = static_cast<int>(airties::eAirtiesTlVId::AIRTIES_ETHERNET_STATS);
+    tlvAirtiesEthStats->supported_extra_stats() = set_supp_stats_val();
+
+    auto lan_ifaces = beerocks::net::network_utils::linux_get_lan_interfaces();
+    for (auto &lan_iface : lan_ifaces) {
+        auto port_list = tlvAirtiesEthStats->create_port_list();
+
+        port_list->port_id() = airties::tlvf_airties_utils::assign_unique_port_id(lan_iface);
+
+        beerocks::net::network_utils::sNetDevStats netDevStats = {0};
+        if (!beerocks::net::network_utils::linux_get_net_dev_stats(lan_iface, netDevStats)) {
+            LOG(WARNING) << "Failed to get net dev stats for " << lan_iface;
+            continue;
+        }
+
+        /*
+         * As per the requirement, only byte counters need
+         * to be converted to KiloBytes.
+         *
+         * The commented out items below correspond to counters that we currently
+         * cannot obtain from the platform (e.g. multicast/broadcast bytes).
+         * These fields are optional, and we keep them commented out for now.
+         * If we later find a reliable source for these counters under /sys or
+         * elsewhere, they can simply be enabled. The insertion order must remain
+         * unchanged.
+         */
+        insert_ethernet_stats_item(port_list, convertBytes_to_Kb(netDevStats.tx_bytes));
+        insert_ethernet_stats_item(port_list, convertBytes_to_Kb(netDevStats.rx_bytes));
+        insert_ethernet_stats_item(port_list, netDevStats.tx_packets);
+        insert_ethernet_stats_item(port_list, netDevStats.rx_packets);
+        insert_ethernet_stats_item(port_list, netDevStats.tx_errs);
+        insert_ethernet_stats_item(port_list, netDevStats.rx_errs);
+        //insert_ethernet_stats_item(port_list, multicast_bytes_sent);
+        //insert_ethernet_stats_item(port_list, multicast_bytes_received);
+        //insert_ethernet_stats_item(port_list, multicast_packets_sent);
+        insert_ethernet_stats_item(port_list, netDevStats.rx_multicast);
+        //insert_ethernet_stats_item(port_list, broadcast_bytes_sent);
+        //insert_ethernet_stats_item(port_list, broadcast_bytes_received);
+        //insert_ethernet_stats_item(port_list, broadcast_packets_sent);
+        //insert_ethernet_stats_item(port_list, broadcast_packets_received);
+
+        tlvAirtiesEthStats->add_port_list(port_list);
+    }
+    LOG(INFO) << "Added Ethernet stats TLV";
+    return true;
+}
+
+/**
+ * @brief Check if the Spanning Tree Protocol (STP) is enabled.
+ *
+ * This function opens a raw socket to communicate with the network device,
+ * then queries the bridge information to determine if STP is enabled.
+ *
+ * @return Returns 1 if STP is enabled, 0 otherwise.
+ */
+bool tlvf_airties_utils::is_airties_platform_common_stp_enabled() const
+{
+    std::string slave_config_file_path =
+        CONF_FILES_WRITABLE_PATH + std::string(BEEROCKS_AGENT) +
+        ".conf"; //search first in platform-specific default directory
+    beerocks::config_file::sConfigSlave beerocks_slave_conf;
+    if (!beerocks::config_file::read_slave_config_file(slave_config_file_path,
+                                                       beerocks_slave_conf)) {
+        slave_config_file_path = mapf::utils::get_install_path() + "config/" +
+                                 std::string(BEEROCKS_AGENT) +
+                                 ".conf"; // if not found, search in beerocks path
+        if (!beerocks::config_file::read_slave_config_file(slave_config_file_path,
+                                                           beerocks_slave_conf)) {
+            std::cout << "config file '" << slave_config_file_path << "' args error." << std::endl;
+            return 0;
+        }
+    }
+    return beerocks::bpl::utils::is_stp_enabled(beerocks_slave_conf.bridge_iface) ? true : false;
+}
+
+constexpr airties::eAirtiesFeatureVersion get_feature_version(eAirtiesFeatureIDs feature_id)
+{
+    switch (feature_id) {
+    /* Eth stats gets version 2 since the introduction of the ETH_STATS_V2 TLV */
+    case eAirtiesFeatureIDs::AIRTIES_FEATURE_ETH_STATS:
+        return airties::eAirtiesFeatureVersion::FEATURE_VERSION_2;
+    default:
+        return airties::eAirtiesFeatureVersion::FEATURE_VERSION_1;
+    }
+}
+
+/**
+ * @brief Create a feature list entry and add it to the TLV.
+ *
+ * This function creates a new feature list entry with the specified feature ID,
+ * sets its version, and adds it to the TLV.
+ *
+ * @param tlvVersionReporting The TLV to which the feature will be added.
+ * @param featureId The ID of the feature to add.
+ * @return void
+ */
+inline void
+create_and_add_feature_to_list(std::shared_ptr<airties::tlvVersionReporting> tlv_version_reporting,
+                               airties::eAirtiesFeatureIDs feature_id)
+{
+
+    // Create a new feature list entry
+    auto version_members = tlv_version_reporting->create_em_agent_feature_list();
+
+    uint16_t version = get_feature_version(feature_id);
+
+    // Set the feature info by combining the feature ID and version
+    // EM+ features supported shall be reported as a big endian 4-octet value, where the 2 lowest
+    // octets shall represent the version (iteration) of a feature and where the
+    // 2 highest octets shall represent the ID of a feature
+    // Below is the value for 2 highest octets of EM+ features supported feature ID.
+    // shifting 16 bits(2 octets) and combining with the feature version.
+    version_members->feature_info() = (static_cast<uint32_t>(feature_id) << 16) | version;
+
+    // Add the created feature list entry to the TLV
+    tlv_version_reporting->add_em_agent_feature_list(version_members);
+}
+
+/**
+ * @brief Add Airties Version Reporting TLV to the CMDU message.
+ *
+ * This function constructs a Version Reporting TLV, populates it with
+ * supported features, and adds it to the outgoing CMDU message.
+ *
+ * @param cmdu_tx The CMDU message to which the TLV will be added.
+ * @return Returns true if the TLV was successfully added, false otherwise.
+ */
+bool tlvf_airties_utils::add_airties_version_reporting_tlv(ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    // Instance to utilize platform-specific utilities
+    airties::tlvf_airties_utils utils_instance;
+
+    // Attempt to create a TLV for version reporting
+    auto tlv_version_reporting = cmdu_tx.addClass<airties::tlvVersionReporting>();
+
+    // Check if the TLV creation failed
+    if (!tlv_version_reporting) {
+        LOG(ERROR) << "Failed to create Airties Feature Profile TLV";
+        return false;
+    }
+
+    // Set the vendor OUI and TLV ID for Airties
+    tlv_version_reporting->vendor_oui() =
+        (sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES));
+
+    tlv_version_reporting->tlv_id() =
+        static_cast<int>(airties::eAirtiesTlVId::AIRTIES_FEATURE_PROFILE);
+
+    // Set the em agent version
+    // EM+ features supported shall be reported as a big endian 4-octet value, where the 2 lowest
+    // octets shall represent the version (iteration) of a feature and where the
+    // 2 highest octets shall represent the ID of a feature
+    // Below is the value for 2 lowest octets of EM+ features supported version.
+    // shifting 16 bits(2 octets) and combining with the subversion.
+    tlv_version_reporting->em_agent_version() =
+        (airties::eMasterVersion::MASTER_VERSION << 16) | airties::eSubVersion::SUB_VERSION;
+
+    // The first feature ID we want to process
+    int count = static_cast<int>(airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_DEVICE_METRICS);
+
+    // Loop through all features until we reach the last one (AIRTIES_FEATURE_END)
+    for (auto feature_id = count;
+         feature_id < static_cast<int>(airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_END);) {
+
+        // Convert integer ID to enum
+        airties::eAirtiesFeatureIDs feature_id_enum =
+            static_cast<airties::eAirtiesFeatureIDs>(feature_id);
+
+        // Handle each feature based on its ID
+        switch (feature_id_enum) {
+
+        // Standard features - always added
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_DEVICE_METRICS:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_IEEE1905_1_14:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_DEVICE_INFO:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_ETH_STATS:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_REBOOT_RESET:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_WIFI6_CAP:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_DPP_ONBOARD:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_SERVICE_STATUS_WIFI_ON_OFF:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_HIDDEN_SSID:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_RADIO_CAPABILITY:
+            // Create and add the feature to the TLV
+            create_and_add_feature_to_list(tlv_version_reporting, feature_id_enum);
+            break;
+
+        // Special case: STP feature is added only if STP is enabled on the platform
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_STP: {
+            if (utils_instance.is_airties_platform_common_stp_enabled()) {
+                LOG(INFO) << "Airties Feature STP is enabled";
+                create_and_add_feature_to_list(tlv_version_reporting, feature_id_enum);
+            }
+            break;
+        }
+        // Not supported
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_LED:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_AFC_SUPPORT:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_TUNNELED_PROBE_REQUEST:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_WPA3:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_WPA_ENTERPRISE:
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_CLIENT_ISOLATION:
+            break;
+        case airties::eAirtiesFeatureIDs::AIRTIES_FEATURE_END: {
+            LOG(INFO) << "Airties Feature END is reached, "
+                      << "Nothing to add in airties-specific version reporting TLV";
+            break;
+        }
+        default: {
+            LOG(WARNING) << "Unknown/unsupported Airties feature ID encountered: "
+                         << static_cast<int>(feature_id_enum) << " (skipping)";
+            break;
+        }
+        }
+        feature_id++;
+    }
+    LOG(INFO) << "Added the airties-specific version reporting TLV";
+    return true;
+}
+
+/*
+ * Function to update the Client details.
+ */
+void update_client_details(std::shared_ptr<airties::tlvAirtiesDeviceInfo> &tlvDevinfo)
+{
+    if (!tlvDevinfo) {
+        LOG(ERROR) << "tlvDevinfo is null";
+        return;
+    }
+
+    static constexpr const char* kDefaultClientId     = "airties_123";
+    static constexpr const char* kDefaultClientSecret = "airties_secret";
+
+    std::string client_id;
+    std::string client_secret;
+
+    // Get client_id and client_password from BPL configuration
+    const bool is_client_id_present = beerocks::bpl::cfg_get_client_id(client_id);
+    const bool is_client_secret_present = beerocks::bpl::cfg_get_client_password(client_secret);
+
+    if (!is_client_id_present || client_id.empty()) {
+        LOG(WARNING) << "BPL: missing or empty client_id in configuration, using default";
+        client_id = kDefaultClientId;
+    }
+    if (!is_client_secret_present || client_secret.empty()) {
+        LOG(WARNING) << "BPL: missing or empty client_password in configuration, using default";
+        client_secret = kDefaultClientSecret;
+    }
+
+    tlvDevinfo->set_client_id(client_id);
+    tlvDevinfo->set_client_secret(client_secret);
+}
+
+/*
+ * Function to add the Device Info TLV
+ * to AP Autoconfiguration WSC Message
+ */
+bool tlvf_airties_utils::add_airties_deviceinfo_tlv(ieee1905_1::CmduMessageTx &m_cmdu_tx)
+{
+    uint32_t randomBootid;
+
+    srand((unsigned)time(NULL));
+    randomBootid = rand();
+
+    auto tlvAirtiesDeviceInfo = m_cmdu_tx.addClass<airties::tlvAirtiesDeviceInfo>();
+    if (!tlvAirtiesDeviceInfo) {
+        LOG(ERROR) << "addClass wfa_map::tlvDeviceInfo failed";
+        return false;
+    }
+    tlvAirtiesDeviceInfo->vendor_oui() =
+        (sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES));
+    tlvAirtiesDeviceInfo->tlv_id()  = static_cast<int>(airties::eAirtiesTlVId::AIRTIES_DEVICE_INFO);
+    tlvAirtiesDeviceInfo->boot_id() = randomBootid;
+
+    update_client_details(tlvAirtiesDeviceInfo);
+
+    bool local_gw = false;
+    {
+        local_gw = beerocks::AgentDB::get()->device_conf.local_gw;
+    }
+
+    if (local_gw) { //it's a controller
+        tlvAirtiesDeviceInfo->flags1().gateway_product_class  = TLV_BIT_ENABLE;
+        tlvAirtiesDeviceInfo->flags2().device_role_indication = TLV_BIT_ENABLE;
+    } else {
+        tlvAirtiesDeviceInfo->flags1().extender_product_class = TLV_BIT_ENABLE;
+        tlvAirtiesDeviceInfo->flags2().device_role_indication = TLV_BIT_DISABLE;
+    }
+    LOG(INFO) << "Added Device Info TLV";
+    return true;
+}
+
+//Global declarations for keeping previous cpu values and meminfo
+uint32_t cpu_idle_prev = 0, cpu_total_prev = 1;
+
+//Function to get the CPU temperature
+bool devicemetrics_get_cpu_temp(uint8_t &cpu_temp)
+{
+    char buf[32] = {0};
+    /* Open ACPI thermal zone sysfs file to read temperature. */
+    FILE *fd = fopen(CPU_TEMP_FILE, "r");
+    if (fd == NULL) {
+        LOG(ERROR) << "cannot open file" << CPU_TEMP_FILE;
+        return false;
+    }
+
+    if (fgets(buf, sizeof(buf), fd)) {
+        /* Temperature resides as milidegree Celcius as denoted in Linux ACPI docs. */
+        cpu_temp = atoi(buf) / 1000;
+    } else {
+        cpu_temp = 0;
+    }
+    fclose(fd);
+    return true;
+}
+
+//Function to get the CPU Load
+bool devicemetrics_get_cpu_load(uint8_t &cpu_load)
+{
+    uint8_t cpu_total = 0, cpu_idle = 0;
+    char buf[256] = {0};
+
+    FILE *fp = fopen(STAT_FILE, "r");
+    if (fp == NULL) {
+        LOG(ERROR) << "cannot open file" << STAT_FILE;
+        return false;
+    }
+
+    /* Get the first line of CPU stats. */
+    if (fgets(buf, sizeof(buf), fp)) {
+        int i = 0;
+        char *token, *ctx;
+
+        /* Check if we have expected string in read buffer. */
+        if (strncmp(buf, STAT_CPU_TXT, strlen(STAT_CPU_TXT)) != 0) {
+            LOG(ERROR) << "Incorrect string read in CPU stats.";
+            fclose(fp);
+            return false;
+        }
+        /* Point empty spaces and parse to get CPU stats: user nice system idle .. */
+        token = strtok_r(buf, " ", &ctx);
+        while (token != NULL) {
+            token = strtok_r(NULL, " ", &ctx);
+            if (token != NULL) {
+                cpu_total += atoi(token);
+                /* IDLE ticks are stored in 4th column according to Linux documentation. */
+                if (i == STAT_IDLE_IND) {
+                    cpu_idle = atoi(token);
+                }
+                i++;
+            }
+        }
+    }
+
+    LOG(INFO) << "cpu_idle_prev " << cpu_idle_prev << "cpu_idle " << cpu_idle << "cpu_total_prev "
+              << cpu_total_prev << "cpu_total " << cpu_total;
+
+    if (cpu_total > 0 && cpu_idle > 0 && cpu_total != cpu_total_prev) {
+        cpu_load = (1 - ((double)(cpu_idle - cpu_idle_prev) / (cpu_total - cpu_total_prev))) * 100;
+    }
+    /* Keep previous data to get delta between cpu load changes. */
+    cpu_idle_prev  = cpu_idle;
+    cpu_total_prev = cpu_total;
+    fclose(fp);
+    return true;
+}
+
+bool devicemetrics_get_meminfo(int32_t &memtotal, int32_t &memfree, int32_t &memcached)
+{
+    FILE *fp        = NULL;
+    char buf[256]    = {0};
+    int32_t membufs = -1;
+
+    fp = fopen(MEMINFO_FILE, "r");
+    if (fp == NULL) {
+        LOG(ERROR) << "cannot open file" << MEMINFO_FILE;
+        goto out;
+    }
+    /* Read meminfo file line by line to fetch free, cached and total sizes. */
+    while (fgets(buf, sizeof(buf), fp) != NULL) {
+        char *ctx = NULL, *fld = NULL, *val = NULL;
+        fld = strtok_r(buf, " ", &ctx);
+        if (fld == NULL) {
+            continue;
+        }
+        /* Point empty spaces and parse to get mem info. */
+        val = strtok_r(NULL, " ", &ctx);
+        if (val == NULL) {
+            continue;
+        }
+        if (strcmp(fld, MEMCACHED_TXT) == 0) {
+            memcached = atoi(val);
+        } else if (strcmp(fld, MEMFREE_TXT) == 0) {
+            memfree = atoi(val);
+        } else if (strcmp(fld, MEMTOTAL_TXT) == 0) {
+            memtotal = atoi(val);
+        } else if (strcmp(fld, MEMBUFFER_TXT) == 0) {
+            membufs = atoi(val);
+        }
+        if (memcached >= 0 && memfree >= 0 && memtotal >= 0 && membufs >= 0) {
+            /* We got all we need, break the loop. */
+            break;
+        }
+    }
+    if ((memcached < 0) || (memtotal < 0) || (memfree < 0) || (membufs < 0)) {
+        LOG(INFO) << "Failed to read meminfo fields memcache: " << memcached
+                  << "memtotal:  " << memtotal << "memfree: " << memfree << "membuffs: " << membufs;
+        goto out;
+    }
+    memcached = memcached + membufs;
+    fclose(fp);
+    return true;
+out:
+    if (fp != NULL) {
+        fclose(fp);
+    }
+    return false;
+}
+
+//Function to get the Device Uptime
+bool devicemetrics_get_uptime(struct timespec &ts)
+{
+    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
+        return false;
+    }
+    return true;
+}
+
+uint8_t get_temp_degC(const char *path)
+{
+    FILE *fd = fopen(path, "r");
+    if (!fd) {
+        return 0;   // cannot open file → return 0
+    }
+
+    long milli = 0;
+    if (fscanf(fd, "%ld", &milli) != 1) {
+        fclose(fd);
+        return 0;   // cannot read → return 0
+    }
+
+    fclose(fd);
+
+    if (milli < 0) {
+        milli = 0;  // safety for invalid negative values
+    }
+
+    // Convert millidegree → °C
+    return (uint8_t)(milli / 1000);
+}
+
+bool devicemetrics_get_radio_info(std::shared_ptr<airties::tlvAirtiesDeviceMetrics> &tlvDevMetrics)
+{
+    auto db = AgentDB::get();
+    const auto &radios = db->get_radios_list();
+
+    for (const auto &radio : radios) {
+        if (radio) {
+            LOG(INFO) << "Radio MAC: " << radio->front.iface_mac << std::endl;
+            LOG(INFO) << "Interface: " << radio->front.iface_name << std::endl;
+            auto rad_list = tlvDevMetrics->create_radio_list();
+            // When either condition is true the buffer is full; break immediately because no
+            // subsequent radio entry will fit either.
+            if (!rad_list || !rad_list->isInitialized()) {
+                LOG(ERROR) << "create_radio_list() failed for radio "
+                           << radio->front.iface_mac
+                           << " - TLV buffer full, skipping remaining radios";
+                break;
+            }
+            rad_list->radio_id() = radio->front.iface_mac;
+            // Reading a single temp1_input for phy00 since all hwmon entries map to same PHY.
+            // Future: when hwmon→radio mapping becomes available, update to read per‑radio temps.
+            rad_list->radio_temperature() = get_temp_degC("/sys/devices/platform/soc@0/c000000.wifi/ieee80211/phy00/hwmon4/temp1_input");
+            tlvDevMetrics->add_radio_list(rad_list);
+        }
+    }
+    LOG(INFO) <<  "No of radios : " << tlvDevMetrics->num_of_radios();
+    return true;
+
+}
+
+/*
+ * Function to add Device Metrics TLV
+ * to AP Metrics Response Message
+ */
+bool tlvf_airties_utils::add_device_metrics(ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    struct timespec ts = {.tv_sec = 0, .tv_nsec = 0};
+    int32_t memcached = -1, memtotal = -1, memfree = -1;
+    uint8_t cpu_load = 0, cpu_temp = 0;
+
+    auto tlvAirtiesDeviceMetrics = cmdu_tx.addClass<airties::tlvAirtiesDeviceMetrics>();
+    if (!tlvAirtiesDeviceMetrics) {
+        LOG(ERROR) << "Failed adding tlvAirtiesDeviceMetrics";
+        return false;
+    }
+
+    tlvAirtiesDeviceMetrics->vendor_oui() =
+        sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES);
+    tlvAirtiesDeviceMetrics->tlv_id() =
+        static_cast<int>(airties::eAirtiesTlVId::AIRTIES_DEVICE_METRICS);
+
+    /*
+     * If any of the following fields like cpu load, meminfo returns
+     * error, the TLV will still be added except the respective values
+     * which threw error.
+     */
+    if (devicemetrics_get_uptime(ts)) {
+        tlvAirtiesDeviceMetrics->uptime_to_boot() = ts.tv_sec;
+    } else {
+        LOG(INFO) << "Unable to fetch the clock time for"
+                  << "updating the Device Metrics TLV";
+        tlvAirtiesDeviceMetrics->uptime_to_boot() = 0;
+    }
+
+    if (devicemetrics_get_cpu_load(cpu_load)) {
+        tlvAirtiesDeviceMetrics->cpu_loadtime_platform() = cpu_load;
+    } else {
+        LOG(INFO) << "Unable to fetch the CPU load for"
+                  << "updating the Device Metrics TLV";
+        tlvAirtiesDeviceMetrics->cpu_loadtime_platform() = 0;
+    }
+
+    if (devicemetrics_get_cpu_temp(cpu_temp)) {
+        tlvAirtiesDeviceMetrics->cpu_temperature() = cpu_temp;
+    } else {
+        LOG(INFO) << "Unable to fetch the CPU Temp for"
+                  << "updating the Device Metrics TLV";
+        tlvAirtiesDeviceMetrics->cpu_temperature() = 0;
+    }
+
+    if (devicemetrics_get_meminfo(memtotal, memfree, memcached)) {
+        tlvAirtiesDeviceMetrics->platform_totalmemory()  = memtotal;
+        tlvAirtiesDeviceMetrics->platform_freememory()   = memfree;
+        tlvAirtiesDeviceMetrics->platform_cachedmemory() = memcached;
+    } else {
+        LOG(INFO) << "Unable to fetch the Memory Info for"
+                  << "updating the Device Metrics TLV";
+        tlvAirtiesDeviceMetrics->platform_totalmemory()  = 0;
+        tlvAirtiesDeviceMetrics->platform_freememory()   = 0;
+        tlvAirtiesDeviceMetrics->platform_cachedmemory() = 0;
+    }
+    if (!devicemetrics_get_radio_info(tlvAirtiesDeviceMetrics)) {
+        LOG(INFO) << "Unable to fetch the radio Info for"
+                  << "updating the Device Metrics TLV";
+    }
+    LOG(INFO) << "Added Device Metrics TLV";
+    return true;
+}
+
+/**
+ * @brief Prototype function to add an Airties Message Type TLV to the CMDU.
+ *
+ * This function demonstrates how to add another TLV (in this case, a Message Type TLV)
+ * to the CMDU message.
+ *
+ * @param cmdu_tx The CMDU message to which the TLV will be added.
+ * @return Returns true if the TLV was successfully added, false otherwise.
+ */
+bool tlvf_airties_utils::add_airties_msgtype_tlv(ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    // Attempt to create a TLV for Airties message type
+    auto tlv_airties_msg_type = cmdu_tx.addClass<airties::tlvAirtiesMsgType>();
+
+    // Check if the TLV creation failed
+    if (!tlv_airties_msg_type) {
+        LOG(ERROR) << "addClass wfa_map::tlvMsgType failed";
+        return false;
+    }
+
+    // Set the vendor OUI for Airties
+    tlv_airties_msg_type->vendor_oui() =
+        (sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES));
+    LOG(INFO) << "Added Airties Msg Type TLV";
+    return true;
+}
+
+/**
+ * @brief Add Airties Radio Capability TLV to the CMDU message.
+ *
+ * This function constructs a Radio Capability TLV, populates it with
+ * supported standards from the radio's actual capabilities (HT/VHT/HE/WiFi6/EHT),
+ * and adds it to the outgoing CMDU message.
+ *
+ * @param cmdu_tx The CMDU message to which the TLV will be added.
+ * @return Returns true if the TLV was successfully added, false otherwise.
+ */
+bool tlvf_airties_utils::add_radio_capability(ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    auto db = beerocks::AgentDB::get();
+    const auto &radios = db->get_radios_list();
+
+    for (const auto &radio : radios) {
+        if (!radio) {
+            continue;
+        }
+
+        auto tlvAirtiesRadioCapability = cmdu_tx.addClass<airties::tlvAirtiesRadioCapability>();
+        if (!tlvAirtiesRadioCapability) {
+            LOG(ERROR) << "addClass airties::tlvAirtiesRadioCapability failed";
+            return false;
+        }
+
+        tlvAirtiesRadioCapability->vendor_oui() =
+            (sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES));
+        tlvAirtiesRadioCapability->tlv_id() =
+            static_cast<int>(airties::eAirtiesTlVId::AIRTIES_RADIO_CAPABILITY);
+
+        tlvAirtiesRadioCapability->radio_id() = radio->front.iface_mac;
+
+        // Initialize all standards to 0
+        tlvAirtiesRadioCapability->standards().s_80211a = 0;
+        tlvAirtiesRadioCapability->standards().s_80211b = 0;
+        tlvAirtiesRadioCapability->standards().s_80211g = 0;
+        tlvAirtiesRadioCapability->standards().s_80211n = 0;
+        tlvAirtiesRadioCapability->standards().s_80211ac = 0;
+        tlvAirtiesRadioCapability->standards().s_80211ax = 0;
+        tlvAirtiesRadioCapability->standards().s_80211be = 0;
+        tlvAirtiesRadioCapability->standards().s_80211bn = 0;
+
+        // Determine supported standards based on frequency band and actual radio capabilities
+        // The radio_info structure is populated from OneWifi data model in refresh_radio_info()
+        
+        // Set basic standards based on frequency band
+        if (radio->frequency_band == beerocks::eFreqType::FREQ_24G) {
+            // 2.4GHz band supports b/g
+            tlvAirtiesRadioCapability->standards().s_80211b = 1;
+            tlvAirtiesRadioCapability->standards().s_80211g = 1;
+        } else if (radio->frequency_band == beerocks::eFreqType::FREQ_5G) {
+            // 5GHz band supports a
+            tlvAirtiesRadioCapability->standards().s_80211a = 1;
+        } else if (radio->frequency_band == beerocks::eFreqType::FREQ_6G) {
+            // 6GHz band - no legacy standards
+        }
+
+        // Check HT (802.11n) support from radio capabilities
+        if (radio->ht_supported) {
+            tlvAirtiesRadioCapability->standards().s_80211n = 1;
+        }
+
+        // Check VHT (802.11ac) support from radio capabilities
+        if (radio->vht_supported) {
+            tlvAirtiesRadioCapability->standards().s_80211ac = 1;
+        }
+
+        // Check HE (802.11ax/WiFi 6) support from radio capabilities
+        if (radio->he_supported) {
+            tlvAirtiesRadioCapability->standards().s_80211ax = 1;
+        }
+
+        // Check EHT (802.11be/WiFi 7) support from radio capabilities
+        if (radio->eht_supported) {
+            tlvAirtiesRadioCapability->standards().s_80211be = 1;
+        }
+
+#ifdef IEEE80211_11BN
+        // Force enable WiFi 7 and WiFi 8 for testing on platforms that support it
+        // This can be removed once OneWifi properly populates radio capabilities
+        tlvAirtiesRadioCapability->standards().s_80211be = 1;
+        tlvAirtiesRadioCapability->standards().s_80211bn = 1;
+#endif
+
+        LOG(INFO) << "Added Radio Capability TLV for radio " << radio->front.iface_mac
+                  << " (iface=" << radio->front.iface_name
+                  << ", freq_type=" << static_cast<int>(radio->frequency_band)
+                  << ", ht=" << radio->ht_supported
+                  << ", vht=" << radio->vht_supported
+                  << ", he=" << radio->he_supported
+                  << ", eht=" << radio->eht_supported
+                  << ", standards: a=" << (int)tlvAirtiesRadioCapability->standards().s_80211a
+                  << " b=" << (int)tlvAirtiesRadioCapability->standards().s_80211b
+                  << " g=" << (int)tlvAirtiesRadioCapability->standards().s_80211g
+                  << " n=" << (int)tlvAirtiesRadioCapability->standards().s_80211n
+                  << " ac=" << (int)tlvAirtiesRadioCapability->standards().s_80211ac
+                  << " ax=" << (int)tlvAirtiesRadioCapability->standards().s_80211ax
+                  << " be=" << (int)tlvAirtiesRadioCapability->standards().s_80211be << ")";
+    }
+
+    return true;
+}
+
+uint8_t tlvf_airties_utils::assign_unique_port_id(const std::string &interface_name)
+{
+    static std::mutex mtx;
+    std::lock_guard<std::mutex> lock(mtx);
+
+    static std::map<std::string, uint8_t> iface_to_id;
+
+    // If already assigned, return existing ID
+    auto it = iface_to_id.find(interface_name);
+    if (it != iface_to_id.end()) {
+        return it->second;
+    }
+
+    // Assign next sequential ID starting from 1
+    uint8_t new_id              = static_cast<uint8_t>(iface_to_id.size() + 1);
+    iface_to_id[interface_name] = new_id;
+
+    return new_id;
+}
+
+/*
+ * Function to add the Ethernet 1905 Neighbor Device List TLV
+ * to Topology Response Message
+ */
+bool tlvf_airties_utils::add_airties_ethernet_1905_neighbor_device_list_tlv(ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    // Use system functions to get LAN interfaces instead of manual getifaddrs()
+    auto lan_ifaces = beerocks::net::network_utils::linux_get_lan_interfaces();
+
+    // First collect all ports that actually have 1905 neighbors
+    struct sPortNeighbors {
+        std::string ifname;
+        uint8_t port_id;
+        std::vector<sMacAddr> al_neighbors;
+    };
+    std::vector<sPortNeighbors> ports_with_neighbors;
+
+    for (const auto &ifname : lan_ifaces) {
+        std::string iface_mac_str;
+        sMacAddr local_iface_mac = {};
+        std::vector<sMacAddr> al_neighbors;
+
+        if (beerocks::net::network_utils::linux_iface_get_mac(ifname, iface_mac_str)) {
+            local_iface_mac = tlvf::mac_from_string(iface_mac_str);
+
+            // Query AgentDB for 1905 AL neighbors behind this interface
+            auto db = beerocks::AgentDB::get();
+            auto it = db->neighbor_devices.find(local_iface_mac);
+            if (it != db->neighbor_devices.end()) {
+                const auto &neighbors_by_al_mac = it->second;
+                al_neighbors.reserve(neighbors_by_al_mac.size());
+                for (const auto &kv : neighbors_by_al_mac) {
+                    al_neighbors.push_back(kv.first);
+                }
+            }
+        }
+
+        if (al_neighbors.empty()) {
+            continue;
+        }
+
+        sPortNeighbors pn;
+        pn.ifname      = ifname;
+        pn.port_id     = airties::tlvf_airties_utils::assign_unique_port_id(ifname);
+        pn.al_neighbors = std::move(al_neighbors);
+        ports_with_neighbors.push_back(std::move(pn));
+    }
+
+    // If no ports have neighbors, skip emitting the TLV to avoid confusing strict controllers.
+    if (ports_with_neighbors.empty()) {
+        LOG(INFO) << "[TLV:1905-NEIGH] No Ethernet ports with 1905 neighbors, skipping TLV";
+        return true;
+    }
+
+    // Create and initialize the 1905 neighbor device list TLV
+    auto tlvNeighborList = cmdu_tx.addClass<airties::tlvAirtiesEth1905NeighborDeviceList>();
+    if (!tlvNeighborList) {
+        LOG(ERROR) << "[TLV:1905-NEIGH] addClass tlvAirtiesEth1905NeighborDeviceList FAILED";
+        return false;
+    }
+
+    tlvNeighborList->vendor_oui() = sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES);
+    tlvNeighborList->tlv_id()     = static_cast<int>(airties::eAirtiesTlVId::AIRTIES_1905_NEIGHBOR_DEVICE_LIST);
+
+    uint8_t added_ports = 0;
+    for (const auto &port_info : ports_with_neighbors) {
+        auto port_entry = tlvNeighborList->create_port_entries();
+        if (!port_entry) {
+            LOG(ERROR) << "[TLV:1905-NEIGH] create_port_entries() FAILED for ifname=" << port_info.ifname;
+            continue;
+        }
+
+        port_entry->port_id() = port_info.port_id;
+
+        const uint8_t num_neighbor =
+            static_cast<uint8_t>(std::min<size_t>(port_info.al_neighbors.size(), 0xFF));
+        port_entry->num_of_neighbor_al_macs() = num_neighbor;
+
+        if (num_neighbor > 0) {
+            if (!port_entry->alloc_neighbor_al_mac_list(num_neighbor)) {
+                LOG(ERROR) << "[TLV:1905-NEIGH] alloc_neighbor_al_mac_list(" << int(num_neighbor)
+                           << ") failed for ifname=" << port_info.ifname;
+                port_entry->num_of_neighbor_al_macs() = 0;
+            } else {
+                for (size_t i = 0; i < num_neighbor; ++i) {
+                    auto tup = port_entry->neighbor_al_mac_list(i);
+                    if (std::get<0>(tup)) {
+                        std::get<1>(tup) = port_info.al_neighbors[i];
+                    }
+                }
+            }
+        }
+
+        tlvNeighborList->add_port_entries(port_entry);
+        ++added_ports;
+    }
+
+    tlvNeighborList->num_of_eth_ports() = added_ports;
+    LOG(INFO) << "Added Ethernet 1905 Neighbor Device List TLV with " << int(added_ports)
+              << " ports";
+    return true;
+}
+
+/*
+ * Function to add the Ethernet Non 1905 Neighbor Device List TLV
+ * to Topology Response Message
+ */
+bool tlvf_airties_utils::add_airties_ethernet_non_1905_neighbor_device_list_tlv(ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    // Use system functions to get LAN interfaces instead of manual getifaddrs()
+    auto lan_ifaces = beerocks::net::network_utils::linux_get_lan_interfaces();
+
+    // Helper functions for FDB processing and MAC filtering
+    auto is_multicast_mac_str = [](const std::string &mac) -> bool {
+        unsigned first_oct = 0;
+        if (mac.size() < 2 || std::sscanf(mac.c_str(), "%2x", &first_oct) != 1) {
+            return true;
+        }
+        return (first_oct & 0x01) != 0;
+    };
+
+    auto get_fdb_unicast_for_dev =
+        [&is_multicast_mac_str](const std::string &dev) -> std::vector<std::string> {
+        std::vector<std::string> macs;
+        FILE *fp = popen("bridge fdb show", "r");
+        if (!fp) {
+            LOG(WARNING) << "[TLV:NON1905-NEIGH] 'bridge fdb show' failed; dev=" << dev;
+            return macs;
+        }
+        char line[512] = {0};
+        while (fgets(line, sizeof(line), fp)) {
+            std::string ln(line);
+            if (ln.find(" dev " + dev) == std::string::npos) {
+                continue;
+            }
+            if (ln.find(" self ") != std::string::npos) {
+                continue;
+            }
+            if (ln.find(" permanent") != std::string::npos) {
+                continue;
+            }
+
+            size_t sp = ln.find(' ');
+            if (sp == std::string::npos) {
+                continue;
+            }
+            std::string mac = ln.substr(0, sp);
+            if (mac.size() != 17) {
+                continue;
+            }
+            if (is_multicast_mac_str(mac)) {
+                continue;
+            }
+
+            if (std::find(macs.begin(), macs.end(), mac) == macs.end()) {
+                macs.push_back(mac);
+            }
+        }
+        pclose(fp);
+        return macs;
+    };
+
+    // First collect only ports that actually have non-1905 neighbors
+    struct sPortNon1905 {
+        std::string ifname;
+        uint8_t port_id;
+        std::vector<sMacAddr> non1905_neighbors;
+    };
+    std::vector<sPortNon1905> ports_with_non1905_neighbors;
+
+    for (const auto &ifname : lan_ifaces) {
+        std::string iface_mac_str;
+        std::vector<sMacAddr> non1905_list;
+
+        if (beerocks::net::network_utils::linux_iface_get_mac(ifname, iface_mac_str)) {
+            sMacAddr local_iface_mac = tlvf::mac_from_string(iface_mac_str);
+            auto db                  = beerocks::AgentDB::get();
+            std::vector<sMacAddr> known_1905_al;
+            auto it = db->neighbor_devices.find(local_iface_mac);
+            if (it != db->neighbor_devices.end()) {
+                const auto &neighbors_by_al = it->second;
+                known_1905_al.reserve(neighbors_by_al.size());
+                for (const auto &kv : neighbors_by_al) {
+                    known_1905_al.push_back(kv.first);
+                }
+            }
+
+            auto fdb_list_str = get_fdb_unicast_for_dev(ifname);
+
+            for (const auto &mac_s : fdb_list_str) {
+                sMacAddr mac = tlvf::mac_from_string(mac_s);
+                bool is_1905 = false;
+                for (const auto &al : known_1905_al) {
+                    if (al == mac) {
+                        is_1905 = true;
+                        break;
+                    }
+                }
+                if (is_1905) {
+                    continue;
+                }
+
+                bool dup = false;
+                for (const auto &m : non1905_list) {
+                    if (m == mac) {
+                        dup = true;
+                        break;
+                    }
+                }
+                if (!dup) {
+                    non1905_list.push_back(mac);
+                }
+            }
+        }
+
+        if (non1905_list.empty()) {
+            continue;
+        }
+
+        sPortNon1905 pn;
+        pn.ifname            = ifname;
+        pn.port_id           = airties::tlvf_airties_utils::assign_unique_port_id(ifname);
+        pn.non1905_neighbors = std::move(non1905_list);
+        ports_with_non1905_neighbors.push_back(std::move(pn));
+    }
+
+    // If no ports have non-1905 neighbors, skip emitting the TLV.
+    if (ports_with_non1905_neighbors.empty()) {
+        LOG(INFO) << "[TLV:NON1905-NEIGH] No Ethernet ports with non-1905 neighbors, skipping TLV";
+        return true;
+    }
+
+    // Create and initialize the non-1905 neighbor device list TLV
+    auto tlvNon1905 = cmdu_tx.addClass<airties::tlvAirtiesEthNon1905NeighborDeviceList>();
+    if (!tlvNon1905) {
+        LOG(ERROR) << "[TLV:NON1905-NEIGH] addClass tlvAirtiesEthNon1905NeighborDeviceList FAILED";
+        return false;
+    }
+
+    tlvNon1905->vendor_oui() =
+        sVendorOUI(airties::tlvAirtiesMsgType::airtiesVendorOUI::OUI_AIRTIES);
+    tlvNon1905->tlv_id() =
+        static_cast<int>(airties::eAirtiesTlVId::AIRTIES_NON_1905_NEIGHBOR_DEVICE_LIST);
+
+    uint8_t added_ports = 0;
+    for (const auto &port_info : ports_with_non1905_neighbors) {
+        auto port_entry = tlvNon1905->create_port_entries();
+        if (!port_entry) {
+            LOG(ERROR) << "[TLV:NON1905-NEIGH] create_port_entries() FAILED for "
+                       << port_info.ifname;
+            continue;
+        }
+
+        port_entry->port_id() = port_info.port_id;
+
+        const uint8_t num_neighbor =
+            static_cast<uint8_t>(std::min<size_t>(port_info.non1905_neighbors.size(), 0xFF));
+        port_entry->num_of_non1905_neighbor_macs() = num_neighbor;
+
+        if (num_neighbor > 0 && !port_entry->alloc_non1905_neighbor_mac_list(num_neighbor)) {
+            LOG(ERROR) << "[TLV:NON1905-NEIGH] alloc_non1905_neighbor_mac_list(" << int(num_neighbor)
+                       << ") failed for " << port_info.ifname;
+            port_entry->num_of_non1905_neighbor_macs() = 0;
+        } else if (num_neighbor > 0) {
+            for (size_t i = 0; i < num_neighbor; ++i) {
+                auto tup = port_entry->non1905_neighbor_mac_list(i);
+                if (std::get<0>(tup)) {
+                    std::get<1>(tup) = port_info.non1905_neighbors[i];
+                }
+            }
+        }
+
+        tlvNon1905->add_port_entries(port_entry);
+        ++added_ports;
+    }
+
+    // Finalize TLV with total port count
+    tlvNon1905->num_of_eth_ports() = added_ports;
+    LOG(INFO) << "Added Ethernet non 1905 Neighbor Device List TLV with " << int(added_ports)
+              << " ports";
+    return true;
+}
diff --git a/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp b/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp
index 84532380e..0f122a954 100644
--- a/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp
+++ b/framework/platform/bpl/cfg/rdk/bpl_cfg.cpp
@@ -1539,5 +1539,78 @@ bool cfg_get_dhcp_gateway_ip(std::string &gw_ip)
     return true;
 }
 
+/**
+ * @brief Helper function to extract a string field from the config JSON file.
+ *
+ * This function reads the config file, parses it as JSON, navigates to root["config"],
+ * and extracts the specified string field.
+ *
+ * @param field_name The name of the field to extract from the config object
+ * @param output The string to store the extracted value
+ * @return true if the field was successfully extracted, false otherwise
+ */
+static bool get_config_string_field(const char *field_name, std::string &output)
+{
+    output.clear();
+
+    if (!config_file_path) {
+        LOG(ERROR) << "config_file_path is null";
+        return false;
+    }
+
+    std::ifstream config_file(config_file_path);
+    if (!config_file.is_open()) {
+        LOG(WARNING) << "Failed to open config file: " << config_file_path;
+        return false;
+    }
+
+    std::string config_content((std::istreambuf_iterator<char>(config_file)),
+                               std::istreambuf_iterator<char>());
+    config_file.close();
+    if (config_file.fail()) {
+        LOG(ERROR) << "unable to close config file: " << config_file_path;
+        return false;
+    }
+
+    cJSON *root = cJSON_Parse(config_content.c_str());
+    if (!root) {
+        const char *error_ptr = cJSON_GetErrorPtr();
+        LOG(ERROR) << "Failed to parse JSON config file: " << config_file_path
+                   << " Error: " << (error_ptr ? error_ptr : "unknown");
+        return false;
+    }
+
+    cJSON *config_obj = cJSON_GetObjectItem(root, "config");
+    if (!config_obj) {
+        LOG(WARNING) << "Missing 'config' object in " << config_file_path;
+        cJSON_Delete(root);
+        return false;
+    }
+
+    cJSON *field_obj = cJSON_GetObjectItem(config_obj, field_name);
+    if (field_obj && cJSON_IsString(field_obj)) {
+        const char *field_str = cJSON_GetStringValue(field_obj);
+        if (field_str && strlen(field_str) > 0) {
+            output = field_str;
+            cJSON_Delete(root);
+            return true;
+        }
+    }
+
+    LOG(WARNING) << "Missing or invalid " << field_name << " in " << config_file_path;
+    cJSON_Delete(root);
+    return false;
+}
+
+bool cfg_get_client_id(std::string &client_id)
+{
+    return get_config_string_field("client_id", client_id);
+}
+
+bool cfg_get_client_password(std::string &client_password)
+{
+    return get_config_string_field("client_password", client_password);
+}
+
 } // namespace bpl
 } // namespace beerocks
diff --git a/framework/platform/bpl/include/bpl/bpl_cfg.h b/framework/platform/bpl/include/bpl/bpl_cfg.h
index c5019503f..3cf5cca95 100644
--- a/framework/platform/bpl/include/bpl/bpl_cfg.h
+++ b/framework/platform/bpl/include/bpl/bpl_cfg.h
@@ -1672,6 +1672,22 @@ bool cfg_set_wifi_iface_multi_ap_vlanid(const std::string &iface_name, uint16_t
  */
 bool cfg_get_dhcp_gateway_ip(std::string &gw_ip);
 
+/**
+ * @brief Get the client ID for Airties device info.
+ *
+ * @param [out] client_id  Client ID string.
+ * @return true on success, otherwise false.
+ */
+bool cfg_get_client_id(std::string &client_id);
+
+/**
+ * @brief Get the client password for Airties device info.
+ *
+ * @param [out] client_password  Client password string.
+ * @return true on success, otherwise false.
+ */
+bool cfg_get_client_password(std::string &client_password);
+
 } // namespace bpl
 } // namespace beerocks
 
diff --git a/framework/tlvf/yaml/tlvf/airties/tlvAirtiesRadioCapability.yaml b/framework/tlvf/yaml/tlvf/airties/tlvAirtiesRadioCapability.yaml
index 4fc2b9a8b..de8a60eac 100644
--- a/framework/tlvf/yaml/tlvf/airties/tlvAirtiesRadioCapability.yaml
+++ b/framework/tlvf/yaml/tlvf/airties/tlvAirtiesRadioCapability.yaml
@@ -44,5 +44,5 @@ sStandards:
     _bit_range: [2,2]
   s_80211be:
     _bit_range: [1,1]
-  reserved:
+  s_80211bn:
     _bit_range: [0,0]
-- 
2.34.1

